CMake/Findmlpack.cmake

```diff
@@ -0,0 +1,641 @@
+      string(REGEX REPLACE ".*#define MLPACK_VERSION_MAJOR ([0-9]+).*" "\\1"
+          MLPACK_VERSION_MAJOR "${_mlpack_HEADER_CONTENTS}")
+      string(REGEX REPLACE ".*#define MLPACK_VERSION_MINOR ([0-9]+).*" "\\1"
+          MLPACK_VERSION_MINOR "${_mlpack_HEADER_CONTENTS}")
+      string(REGEX REPLACE ".*#define MLPACK_VERSION_PATCH ([0-9]+).*" "\\1"
+          MLPACK_VERSION_PATCH "${_mlpack_HEADER_CONTENTS}")
+
+    endif()
+
+    set(MLPACK_VERSION_STRING "${MLPACK_VERSION_MAJOR}.${MLPACK_VERSION_MINOR}.${MLPACK_VERSION_PATCH}")
+  endif()
+
+  include(FindPackageHandleStandardArgs)
+  find_package_handle_standard_args(mlpack
+    REQUIRED_VARS MLPACK_INCLUDE_DIR
+    VERSION_VAR MLPACK_VERSION_STRING)
+
+  mark_as_advanced(MLPACK_INCLUDE_DIR)
+  mark_as_advanced(MLPACK_INCLUDE_DIRS)
+  mark_as_advanced(MLPACK_LIBRARIES)
+
+endmacro()
+
+macro(fetch_mlpack compile)
+  set(version 0.3.28)
+  find_package(BLAS PATHS ${CMAKE_BINARY_DIR})
+  if (NOT BLAS_FOUND OR (NOT BLAS_LIBRARIES))
+    get_deps(https://github.com/xianyi/OpenBLAS/releases/download/v${version}/OpenBLAS-${version}.tar.gz OpenBLAS OpenBLAS-${version}.tar.gz)
+    if (NOT compile)
+      message(WARNING "OpenBLAS is downloaded but not compiled. Please compile
+      OpenBLAS before compiling mlpack")
+    else()
+      execute_process(COMMAND make NO_SHARED=1 WORKING_DIRECTORY ${CMAKE_BINARY_DIR}/deps/OpenBLAS-${version})
+      file(GLOB OPENBLAS_LIBRARIES "${CMAKE_BINARY_DIR}/deps/OpenBLAS-${version}/libopenblas.a")
+      set(BLAS_openblas_LIBRARY ${OPENBLAS_LIBRARIES})
+      set(LAPACK_openblas_LIBRARY ${OPENBLAS_LIBRARIES})
+      set(BLAS_FOUND ON)
+    endif()
+  endif()
+
+  find_armadillo(${CMAKE_BINARY_DIR})
+  if (NOT ARMADILLO_FOUND)
+    if (NOT CMAKE_CROSSCOMPILING)
+      find_package(BLAS QUIET)
+      find_package(LAPACK QUIET)
+      if (NOT BLAS_FOUND AND NOT LAPACK_FOUND)
+        message(FATAL_ERROR "Can not find BLAS or LAPACK!  These are required for Armadillo.  Please install one of them---or install Armadillo---before installing mlpack.")
+      endif()
+    endif()
+    get_deps(https://files.mlpack.org/armadillo-${ARMADILLO_VERSION}.tar.gz armadillo armadillo-${ARMADILLO_VERSION}.tar.gz)
+    set(ARMADILLO_INCLUDE_DIR ${GENERIC_INCLUDE_DIR})
+    if (NOT CMAKE_CROSSCOMPILING)
+      find_armadillo(${CMAKE_BINARY_DIR})
+    endif()
+    # Include directories for the previous dependencies.
+    set(MLPACK_INCLUDE_DIRS ${MLPACK_INCLUDE_DIRS} ${ARMADILLO_INCLUDE_DIRS})
+    set(MLPACK_LIBRARIES ${MLPACK_LIBRARIES} ${ARMADILLO_LIBRARIES})
+  endif()
+
+  find_ensmallen(${CMAKE_BINARY_DIR})
+  if (NOT ENSMALLEN_FOUND)
+    get_deps(https://www.ensmallen.org/files/ensmallen-latest.tar.gz ensmallen ensmallen-latest.tar.gz)
+    set(ENSMALLEN_INCLUDE_DIR ${GENERIC_INCLUDE_DIR})
+    find_ensmallen(${CMAKE_BINARY_DIR})
+    if (ENSMALLEN_FOUND)
+      set(MLPACK_INCLUDE_DIRS ${MLPACK_INCLUDE_DIRS} "${ENSMALLEN_INCLUDE_DIR}")
```

Comment:
The `""`s make me think... variables are awful to deal with when there are spaces in the path, but users might do this.  Can you test that the macros work locally in a build directory with a space?  Something like `mlpack/build test/` will work.  If everything breaks, probably we need more quotes around variables, so, e.g., below, we probably need `set(MLPACK_INCLUDE_DIRS ${MLPACK_INCLUDE_DIRS} "${CEREAL_INCLUDE_DIR}")`.  However, I'm not sure exactly how it has to be... testing locally like I suggested should be sufficient to find whatever issues we have.